KEYCLOAK-9404 Add support for multiple variants of the same social identity provider

thomasdarimont opened the pull request:
This effectively reverts KEYCLOAK-5932 (removal of used social idp)
and allows to specify dedicated aliases and display names for
social IdPs.

This enables users to connect a realm with multiple social IdPs of
the same type, e.g. "google apps", with different configurations.

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
KEYCLOAK-9404 Add support for multiple variants of the same social identity provider

This effectively reverts KEYCLOAK-5932 (removal of used social idp)
and allows to specify dedicated aliases and display names for
social IdPs.

This enables users to connect a realm with multiple social IdPs of
the same type, e.g. "google apps", with different configurations.

abstractj commented:
@ssilvert maybe @vmuzikar has some thoughts about this. But I believe that integration tests should be required to this PR.

vmuzikar reviewed (changes requested):
@thomasdarimont Could you please add some test for this? I think [SocialLoginTest](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/SocialLoginTest.java) would be an appropriate place.

Please double check that all social providers are updated by this. E.g. GitLab uses own template so it's missing the new alias field from the generic template.

I wonder if we should display the alias by default in the IdPs table in Admin Console like we do for non-social IdPs. Currently when Display Name is not specified, it can be a bit confusing:
![Screenshot 2020-06-19 at 09 10 07](https://user-images.githubusercontent.com/13906855/85106588-bd176480-b20c-11ea-863f-24b665172177.png)

Lastly, I'm not sure this really solves KEYCLOAK-9404. Looking at the description the reporter wants to have different set of IdPs available for different clients in the same realm, right? And this we don't support AFAIK.

thomasdarimont committed:
KEYCLOAK-9404 Show alias in the IdP list to differentiate between IdPs of the same type

thomasdarimont commented:
@vmuzikar I gave this another spin and updated the PR

I followed your suggestion and checked if every social IdP shows configurable alias fields.
![image](https://user-images.githubusercontent.com/314690/96651187-8ffece00-1334-11eb-86e7-fbd2ef55f99e.png)

You were spot on, indeed bitbucket and gitlab needed some additional configuration.

![image](https://user-images.githubusercontent.com/314690/96651249-b4f34100-1334-11eb-9c16-db298d7965b2.png)

![image](https://user-images.githubusercontent.com/314690/96651268-bf153f80-1334-11eb-9ed3-258d01184f74.png)

vmuzikar reviewed (changes requested):
@thomasdarimont Thank you for the update! I'm sorry for a very late response from me.

It seems that all IdP now properly have the "Alias" field.

The only thing I'm currently missing is the alias being reflected in the IdPs table as I mentioned in my previous comment.

Also, I'm thinking how to test this. Maybe something like: add two Facebook IdP with different aliases? No need to check the login, just try to add them using the Admin Console. Also, maybe the [Admin Console UI tests module](https://github.com/keycloak/keycloak/tree/master/testsuite/integration-arquillian/tests/other/console) would be most suitable for this.

thomasdarimont commented:
@vmuzikar I just rebased this PR and added the alias to the Identity Provider listing
![image](https://user-images.githubusercontent.com/314690/107220226-60262400-6a12-11eb-8344-ffc0bace6bb6.png)

thomasdarimont commented:
I changed the order from DisplayName, Alias, Prover to DisplayName, Provider, Alias which makes more sense:
`Acme Apps` is a `Google Identity Provider` with alias `google_v2`

vmuzikar commented:
@thomasdarimont Thank you for the update and sorry for late response. 

Alias looks good now but I can see a new UI glitch. The header with "Add provider..." header seems unaligned with the underlaying table.

Also, I'm missing some simple UI test as I mentioned in a previous comment.

ssilvert commented:
@thomasdarimont, I'm going to close this for now because it has become dormant.  If you and @vmuzikar want to continue we can reopen.
